localstack/services/sqs/provider.py

```diff
@@ -671,26 +666,28 @@ def on_before_stop(self):
     def _add_queue(self, queue: SqsQueue, context: RequestContext):
         with self._mutex:
             self.queues[queue.key] = queue
-            self.queue_url_index[queue.url(context)] = queue
-            self.queue_name_index[queue.name] = queue.url(context)
 
-    def _require_queue_by_url(self, queue_url: str) -> SqsQueue:
-        """
-        Returns the queue for the given url, or raises a NonExistentQueue error.
-
-        :param queue_url: The QueueUrl
-        :returns: the queue
-        :raises NonExistentQueue: if the queue does not exist
+    def _require_queue(self, context, queue_url=None, queue_name=None) -> SqsQueue:
         """
+        #     Returns the queue for the given url or name, or raises a NonExistentQueue error.
+        #
+        #     :param queue_url: The QueueUrl
+        #     :param queue_name: The QueueName
+        #     :returns: the queue
+        #     :raises NonExistentQueue: if the queue does not exist
+        #"""
+
+        # if we have nothing, something went probably wrong, and we do a best effort extraction
+        if not (queue_url or queue_name):
+            queue_url = context.request.url
+        if queue_url:
+            queue_name = queue_url.split("/")[-1]
+        key = QueueKey(context.region, context.account_id, queue_name)
```

Comment:
i think it would be better to separate the two aspects of 1) fetching the queue per key and 2) creating the `QueueKey` from the context. my suggestion would be that signature of the `_require_queue` method be `_require_queue(self, key: QueueKey)` and that there's a second method, something like `resolve_queue_key(context, ...`.  then you can chain `_require_queue(resolve_queue_key(...))` which makes it more composable.